src/cargo/core/source/source_id.rs

```diff
@@ -135,6 +135,11 @@ impl SourceId {
                 Ok(SourceId::new(SourceKind::Registry, url, None)?
                     .with_precise(Some("locked".to_string())))
             }
+            "sparse" => {
+                let url = string.into_url()?;
+                Ok(SourceId::new(SourceKind::Registry, url, None)?
+                    .with_precise(Some("locked".to_string())))
```

Comment:
http registries are still `SourceKind::Registry`, so the `sparse+` prefix is how it knows we're in http mode. Otherwise it couldn't tell them apart.

I may be able to make the "registry" one also pass the `registry+` prefix for consistency -- as long as it doesn't show up somewhere external as you mentioned.